### PR TITLE
 feat(gcp): Add Cloud Functions Ingestion and Documentation

### PR DESCRIPTION
**Pull request by janithashri**
### Summary
This Pull Request adds support for ingesting Google Cloud Functions (GCF) data into Cartography. It introduces a new ingestion module that discovers Cloud Functions, maps their key properties, and establishes relationships to their parent GCP Projects.

### Desired Behavior
Upon running a GCP sync, Cloud Functions will be discovered, and new `GCPCloudFunction` nodes will be created in the graph, linked to their respective `GCPProject` nodes via a `[:RESOURCE]` relationship. Key properties such as name, runtime, status, creation/update times, and trigger details (HTTPS URL, event type/resource) will be stored.

### Scenario Enabled
This enables Cartography users to gain comprehensive visibility into their serverless compute resources on Google Cloud. It facilitates asset inventory, aids in security posture assessment (e.g., identifying publicly exposed HTTP functions), and helps understand function dependencies within the overall GCP infrastructure.

### Checklist (Proof of Work)
- [x] **Update/add unit or integration tests.**
- [x] **Include a screenshot showing what the graph looked like before and after your changes.**
- [x] **Include console log trace showing what happened before and after your changes.**

If you are changing a node or relationship:
- [x] **Update the [schema](https://github.com/cartography-cncf/cartography/tree/master/docs/root/modules) and [readme](https://github.com/cartography-cncf/cartography/blob/master/docs/schema/README.md).**
(Schema documentation added directly to `docs/root/modules/gcp/schema.md` and linked from `docs/root/usage/schema.md`.)

If you are implementing a new intel module:
- [x] **Use the NodeSchema [data model](https://cartography-cncf.github.io/cartography/dev/writing-intel-modules.html#defining-a-node).**
(Defined `GCPCloudFunctionNode` schema in `cartography/models/gcp/gcf.py`.)

---

### **Proof of Work:**

Screenshot (409)
Screenshot (408)

**Console Log Trace:**
Screenshot (410)

### CODE REVIEW

File: docs/root/modules/gcp/schema.md [hunk 1]
```diff
@@ -685,3 +685,30 @@ Representation of a GCP [Role](https://cloud.google.com/iam/docs/reference/rest/
     ```
     (GCPRole)-[RESOURCE]->(GCPProject)
     ```
++### GCPCloudFunction
+
+### `GCPCloudFunction`
+Represents a Google Cloud Function.
+
+**Properties:**
```
Comment (achantavy) [inline docs/root/modules/gcp/schema.md]: Please follow the same doc format as the rest of the schema doc so that everything is consistent.

File: cartography/models/gcp/gcf.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    )
+    direction: LinkDirection = LinkDirection.INWARD
+    rel_label: str = "RESOURCE"
+    properties: GCPCloudFunctionToGCPProjectRelProperties = GCPCloudFunctionToGCPProjectRelProperties()
+
+@dataclass(frozen=True)
+class GCPCloudFunctionNode(CartographyNodeSchema):
+    label: str = "GCPCloudFunction"
+    properties: GCPCloudFunctionProperties = GCPCloudFunctionProperties()
+    sub_resource_relationship: Optional[GCPCloudFunctionToGCPProjectRel] = GCPCloudFunctionToGCPProjectRel()
```
Comment (achantavy) [inline cartography/models/gcp/gcf.py]: This formatting is not consistent with the other data models used in cartography. Please view others in cartography/models so that everything is consistent.

**Comment by janithashri** [comment]
hey @achantavy! this PR’s ready for review . I added support for GCP Cloud Functions as discussed.

i’ve cleaned up naming (used snake_case consistently), separated model + intel logic properly, and followed the patterns used across the project (CartographyNodeSchema, TargetNodeMatcher, etc). also added docs and integration test — ran a manual sync and verified it creates the right node + edge (GCPCloudFunction)-[RUNS_AS]->(GCPServiceAccount).

let me know if anything feels off or needs tweaks. thanks a lot for your time!

File: cartography/intel/gcp/gcf.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    project_id: str,
+    update_tag: int,
+) -> None:
+    """
+    Ingests GCP Cloud Functions and their relationships using a single, robust Cypher query.
+    """
+    logger.info(f"Loading {len(data)} GCP Cloud Functions for project {project_id}.")
+
+    ingest_query = """
+    UNWIND $FunctionList as func_data
```
Comment (achantavy) [inline cartography/intel/gcp/gcf.py]: Hey thanks for getting started but please use the data model for this instead of raw Neo4j queries. I see that you have a data model defined already, please make sure to use that here.

Comment (janithashri) [reply cartography/intel/gcp/gcf.py]: I actually started out using the data model with load(...) for this, but ran into some issues as the relationships weren’t resolving correctly in the graph.

I noticed that a most of the other GCP modules like gcp_dns.py and gcp_iam.py also use raw Cypher in similar situations, so I went with that to keep things consistent and to make sure the data linked up properly.

Happy to switch it back to the schema-based approach if needed — just wanted to share the reasoning behind the current setup. Let me know what you think!

Comment (achantavy) [reply cartography/intel/gcp/gcf.py]: Yeah we're working on slowly refactoring things over to the new structure. Thanks for thinking about consistency. Since this is a new module, adding in the data model shouldn't interfere with the older modules. Happy to help on Slack if you have trouble making some of the rels resolve

**Comment by janithashri** [comment]
Hi @achantavy,

Here’s a quick summary of what’s been updated:

Refactored to Schema Model: The ingestion logic now uses the load() function from the schema model in intel/gcp/gcf.py, replacing the earlier raw Cypher queries as you suggested.

Handled Framework Limitation: To work around the issue we encountered with loading complex schemas, I’ve split the process into two separate load() calls. This ensures both the :RESOURCE and :RUNS_AS relationships are correctly established.

Maintained Consistency: The model in models/gcp/gcf.py now reuses the GCPPrincipalToProject relationship and follows the same naming conventions (projectId) as the IAM module.

Improved Testing: Integration tests now thoroughly check for the correct creation and direction of all nodes and relationships. Everything is passing.

Updated Documentation: I’ve added the GCPCloudFunction documentation to schema.md following the project's format.

Final screenshots and logs have also been included in the PR description.

Let me know if there’s anything else you'd like changed. Thanks again for your support!

**Comment by janithashri** [comment]
I have also updated the outdated relationship diagrams for GCPServiceAccount and GCPRole in schema.md to improve clarity and reflect current implementation.

File: cartography/intel/gcp/gcf.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    Ingests GCP Cloud Functions using the Cartography data model.
+    """
+    logger.info(f"Loading {len(data)} GCP Cloud Functions for project {project_id}.")
+
+    transformed_data = []
+    for func_data in data:
+        func_data['https_trigger_url'] = func_data.get('httpsTrigger', {}).get('url')
+        func_data['event_trigger_type'] = func_data.get('eventTrigger', {}).get('eventType')
+        func_data['event_trigger_resource'] = func_data.get('eventTrigger', {}).get('resource')
+        transformed_data.append(func_data)
```
Comment (d-aggarwal) [inline cartography/intel/gcp/gcf.py]: Please consider moving the data transformation logic out of the load function into a separate transform function.

File: tests/integration/cartography/intel/gcp/test_gcf.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+from unittest.mock import patch, MagicMock
+import neo4j
+
+import cartography.intel.gcp.gcf as gcf
+from cartography.client.core.tx import load
+from cartography.models.gcp.iam import GCPServiceAccountSchema
+
+# --- Test Data (Mock GCP API Response) ---
+GCP_FUNCTIONS_RESPONSE = {
+    "functions": [
```
Comment (d-aggarwal) [inline tests/integration/cartography/intel/gcp/test_gcf.py]: Please move test data to a separate file in data/gcp folder

File: tests/integration/cartography/intel/gcp/test_gcf.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        {"UPDATE_TAG": update_tag, "PROJECT_ID": project_id},
+    )
+
+    # Assert 1: Check that the Cloud Function nodes were created.
+    expected_nodes = {
+        "projects/test-project/locations/us-central1/functions/function-1",
+        "projects/test-project/locations/us-east1/functions/function-2",
+    }
+    nodes = neo4j_session.run("MATCH (n:GCPCloudFunction) RETURN n.id")
+    actual_nodes = {n['n.id'] for n in nodes}
```
Comment (d-aggarwal) [inline tests/integration/cartography/intel/gcp/test_gcf.py]: consider using the existing check_nodes and check_rels util functions instead of raw Cypher queries for verifying relationships and node  in integration tests.

File: tests/integration/cartography/intel/gcp/test_gcf.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    # The relationship is (GCPProject)-[:RESOURCE]->(GCPCloudFunction)
+    rels_to_project = neo4j_session.run(
+        """
+        MATCH (p:GCPProject{id:$PROJECT_ID})-[:RESOURCE]->(f:GCPCloudFunction)
+        RETURN count(f) AS rel_count
+        """,
+        PROJECT_ID=project_id,
+    )
+    assert rels_to_project.single()['rel_count'] == 2
+
```
Comment (d-aggarwal) [inline tests/integration/cartography/intel/gcp/test_gcf.py]: Same here for relationships.

File: tests/integration/cartography/intel/gcp/test_gcf.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    # The relationship is (GCPCloudFunction)-[:RUNS_AS]->(GCPServiceAccount)
+    rels_to_sa = neo4j_session.run(
+        """
+        MATCH (f:GCPCloudFunction)-[:RUNS_AS]->(sa:GCPServiceAccount)
+        WHERE f.project_id = $PROJECT_ID
+        RETURN count(sa) as rel_count
+        """,
+        PROJECT_ID=project_id,
+    )
+    assert rels_to_sa.single()['rel_count'] == 2
```
Comment (d-aggarwal) [inline tests/integration/cartography/intel/gcp/test_gcf.py]: Here as well.

**Review by d-aggarwal — CHANGES_REQUESTED** [review]
Thanks for the contribution! I've suggested a few changes to help improve structure and align with new data model. These are mainly to support your understanding and help you move forward while waiting for the final review. Feel free to make these updates in the meantime!

File: tests/integration/cartography/test_basic.py [hunk 9]
```diff
@@ -13,4 +13,4 @@ def test_neo4j_connection():
 def test_create_indexes(neo4j_session):
     # ensure the idempotency of creating indexes
     create_indexes.run(neo4j_session, None)
-    create_indexes.run(neo4j_session, None)
+    create_indexes.run(neo4j_session, None)
```
Comment (achantavy) [inline tests/integration/cartography/test_basic.py]: Removing this empty line will cause the linter to complain; add it back to make it stop complaining

File: cartography/models/gcp/gcf.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        GCPCloudFunctionToGCPServiceAccountRelProperties()
+    )
+
+
+@dataclass(frozen=True)
+class GCPCloudFunctionSchema(CartographyNodeSchema):
+    label: str = "GCPCloudFunction"
+    properties: GCPCloudFunctionProperties = GCPCloudFunctionProperties()
+    # This now reuses the generic relationship from the iam model
+    sub_resource_relationship: GCPPrincipalToProject = GCPPrincipalToProject()
```
Comment (achantavy) [inline cartography/models/gcp/gcf.py]: Don't do this, just create another rel specific for the CloudFunction to project. They would have the same properties, but they are for different objects altogether so as the code changes this could introduce problems.

File: cartography/intel/gcp/gcf.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+) -> None:
+    """
+    Ingests transformed GCP Cloud Functions using the Cartography data model.
+    """
+    logger.info(f"Loading {len(data)} GCP Cloud Functions for project {project_id}.")
+
+    data.sort(key=lambda x: x['name'].split('/')[3] if len(x['name'].split('/')) > 3 else None)
+    for region, functions_in_region in groupby(data, key=lambda x: x['name'].split('/')[3] if len(x['name'].split('/')) > 3 else None):
+        if region:
+            functions_list = list(functions_in_region)
```
Comment (achantavy) [inline cartography/intel/gcp/gcf.py]: This looks terrifying lol.

Can we move it to a transform function so that the `load_*()` function is just a dumb wrapper around `load()`, and can we make this look less scary? haha

File: .gcloudignore [hunk 12]
```diff
@@ -0,0 +1,17 @@
+# This file specifies files that are *not* uploaded to Google Cloud
```
Comment (achantavy) [inline .gcloudignore]: Do we need this file?

**Review by achantavy — CHANGES_REQUESTED** [review]
